### ISSUE TITLE
Architecture

### ISSUE THREAD

**Issue by kieranpotts** [craftingpress/crafting-css#4]
# CSS Architecture

CSS is notoriously difficult to manage in large, complex, rapidly-iterated systems. ... Well-structured CSS yields maintainability and better collaboration, which ultimately deliver better user experiences.

One reason is that CSS lacks a built-in scoping mechanism. Everything in CSS is global. That means any change you make has the potential to cascade and alter the presentation of unrelated bits of the UI.

Extended CSS languages (otherwise known as CSS preprocessors) such as Sass, Less and Stylus make things a little easier by offering up features that make writing CSS easier. But even these solutions don’t truly fix CSS’s scalability issue.

Until CSS gets its own native scoping mechanism, we need to devise our own system for locking down styles to specific sections of an HTML document.

CSS methodologies are the solution.

CSS methodologies are formal, documented systems for authoring CSS in a way that allows us to develop, maintain and scale the front-end as a set of small, isolated modules rather than as one massive lump of indivisible code.

In this article, I take a look at these CSS methodologies:

- Object-Oriented CSS (OOCSS)
- Block, Element, Modifier (BEM)
- Scalable and Modular Architecture for CSS (SMACSS)
- SUIT CSS

Plus I explain the methodology that I deploy in my own projects to keep CSS manageable at scale.

Each CSS methodology offers a slightly different set of solutions to the CSS scalability/maintainability problem. A CSS methodology will often define guidelines for:

- CSS and HTML best practices
- Naming conventions
- Ordering and grouping of CSS style rules
- Code formatting

There is no “best” CSS methodology. Different approaches work better for different individuals/teams/projects.

Hopefully, looking at a few existing CSS methodologies will help you discover one that fits your needs. Or it might inspire you to create your own custom CSS-authoring guidelines.

Adopting a CSS methodology — even if it’s one that you create yourself — will make it easier for you to design and iterate on your web design projects, regardless of scale and complexity.

## [Object-Oriented CSS (OOCSS)](http://oocss.org/)

Nicole Sullivan’s Object-Oriented CSS, or OOCSS for short, was published in 2009.

It was the first CSS methodology to become widely adopted. It’s still hugely influential today.

OOCSS advocates the separation of structure from skin. The methodology makes a clear distinction between content and its containers.

In OOCSS, style rules are written exclusively using CSS class selectors.

For example, the style of button elements might be set via two classes: ``button`` and ``grey-btn``:

```
.button provides the button's basic structure
.grey-btn applies colors and other visual properties
.button {
    box-sizing: border-box;
    height: 50px;
    width: 100%;
}
.grey-btn {
    background: #EEE;
    border: 1px solid #DDD;
    box-shadow: rgba(0, 0, 0, 0.5) 1px 1px 3px;
    color: #555;
}
```

One goal of the OOCSS methodology is to reduce duplication of the same properties throughout your various style rules. In other words, OOCSS can help us maintain DRY stylesheets. The methodology attempts to achieve this goal by using lots of small, modular, specialist CSS classes.

Very few style properties are applied via type selectors (e.g. ``h1``, ``div`` and ``body``).

The use of descendant selectors is discouraged, too.

```
.wrapper .blog-post .button {
    //...
}
```

This way, the presentation of HTML elements is never dependent upon them being used in a specific context or DOM structure.

The great thing about the OOCSS methodology is it urges authors to make the most of CSS’s cascading behavior, rather than trying to lock it up with high-specificity selectors.

Authors are encouraged, whenever possible, to reuse existing style rules instead of creating new ones. We are also encouraged to extend style rules by specifying additional CSS properties with new classes, rather than modifying or overwriting existing CSS properties.

The main downside of OOCSS is that you can end up with an awful lot of classes. Which can be difficult to maintain and manage.

And the concept of object-oriented programming — which is where OOCSS takes inspiration from — does not fit comfortably with CSS.

But that’s not to say the principles of OOCSS are invalid. On the contrary, OOCSS is a commonsense CSS methodology that brings back sanity to large-scale CSS development.

## [Block, Element, Modifier (BEM)](https://en.bem.info/)

Block, Element, Modifier — more commonly called BEM — is a CSS class-naming system devised by the dev team at Yandex (the Google of Russia).

The idea behind BEM is to differentiate CSS classes that fulfill different roles. This is done by naming CSS classes in a way that indicates their role.

BEM complements OOCSS because OOCSS doesn’t impose any particular class-naming convention.

In BEM terminology, a block is an independent, modular UI component. A block may be composed of multiple HTML elements, or even multiple blocks. An example of a block might be your navigation menu or search form.

An element is a component of a block. An element serves a singular purpose. For example, if you have a navigation menu block, then elements of it might be your navigation menu’s links, which in turn might be in the form of list items (``<li>`` elements) and links (``<a>`` elements).

A modifier is a CSS class that changes the default presentation of a block or element.

This is the BEM class-naming syntax:

```
.block
.block--modifier
.block__element
.block__element--modifier
```

Consider the following example, the markup for a login form:

```
<form>
    <label>Username <input type="text" name="username" /></label>
    <label>Password <input type="password" name="password" /></label>
   <button>Sign in</button>
</form>
```

Here’s the markup above with BEM classes applied:

```
<form class="loginform loginform--errors">
    <label class="loginform__username loginform__username--error">
        Username <input type="text" name="username" />
    </label>
    <label class="loginform__password">
        Password <input type="password" name="password" />
    </label>
    <button class="loginform__btn loginform__btn--inactive">
        Sign in
    </button>
</form>
```

The ``.loginform`` class is the block. It is composed of three elements:

- ``.loginform__username``: username input field
- ``.loginform__password``: password input field
- ``.loginform__btn``: form submission button
```

The three modifiers are:

- ``.loginform__username--error``: when there's an error, the element's visual properties are modified so that it indicates to the user there was an error
- ``.loginform__btn--inactive``: the element's visual properties are modified so that it has an inactive appearance
- ``.loginform--errors``: this modifier visually presents the login form in a manner indicating that it contains errors

The BEM naming convention helps CSS authors comply with the OOCSS principle of using a flat selector hierarchy composed of equally-specific class selectors. It also helps OOCSS authors avoid deep descendant selectors.

So instead of the following style rule, which uses three CSS class selectors…

```
.loginform .username .error {
    //...
}
```

… you can just use a single class selector:

```
.loginform__username--error {
    //...
}
```

BEM is a very robust class-naming convention.

It successfully distinguishes the different concerns that classes are used for. And it’s easy to see in the markup which classes are related to one another.

A couple of subjective criticisms against BEM are:

- The class names can end up being long and ugly.
- The naming convention is not intuitive to inexperienced developers.

Personally, I don’t think those are particularly strong criticisms. However, it would be nice to have a class-naming convention that is as robust as BEM, while also being a little tidier and more intuitive.


## [Scalable and Modular Architecture for CSS (SMACSS)](https://smacss.com/)

Jonathan Snook published his book on Scalable and Modular Architecture for CSS in 2011. It is abbreviated as SMACSS and pronounced as “smacks”.

A key idea in this CSS methodology is how we categorize our CSS style rules.

Snook came up with five categories:

- Base styles
- Layout
- Modules
- State
- Themes

Base styles set the default CSS properties of individual HTML elements. These are typically CSS type selectors. Examples:

```
h1 {
   font-size: 32px;
}
div {
    margin: 0 auto;
}
a {
    color: blue;
}
```

Layout rules are related to the structural layout of web pages: containers, the grid, etc. They are prefixed with ``layout-`` or ``l-``. Examples:

```
.layout-sidebar {
    width: 320px;
}
.l-comments {
    width: 640px;
}
```

Modules are style rules that format modular, reusable components. Examples:

```
.call-to-action-button {
    text-transform: uppercase;
    color: #FFF200;
}
.search-form {
    display: inline-block;
    background-color: #E1E1E1;
}
```

State rules are dedicated to specifying the current state of something in the interface. Examples:

```
.is-hidden {
    display: none;
}
.is-highlighted {
    color: #FF0000;
    background-color: #F4F0BB;
    border: 1px solid #CBBD15;
}
```

Finally, themes are style rules that affect layout and modules and are triggered by user preferences/actions/viewing contexts.

SMACSS offers a simpler naming convention than BEM. There are no names for base styles because only type selectors (``h1``, ``p``, ``a``, etc.) are used for those. Modules are given their own unique class names. Sub-components and variations are prefixed with the name of their parent module.

Let’s say our layout container is called ``.l-footer``. We have a search form module inside it. The search form has already been submitted at least once by the user. This might be our markup:

```
<section class="l-footer">
    <form class="search is-submitted">
        <input type="search" />
        <input type="button" value="Search">
    </form>
</section>
```

SMACSS discourages the use of descendant selectors. Jonathan Snook explains this by introducing a concept he calls depth of applicability. This concept is all about limiting the impact that CSS has on HTML markup by using classes to very precisely target elements you want to style.


## [SUIT CSS](http://suitcss.github.io/)

Nicolas Gallagher’s SUIT CSS, introduced in 2014, is interesting because it combines a BEM-like class-naming system with a CSS preprocessor. So SUIT CSS provides us with extended CSS syntax a la Sass, Less or Stylus.

SUIT CSS classes come in five formats:

- ``u-utilityName``
- ``ComponentName``
- ``ComponentName--modifierName``
- ``ComponentName-elementName``
- ``ComponentName.is-stateOfName``

This class-naming convention highlights the division between:

- General utility classes
- Standalone/modular UI components
- Individual elements
- Modifiers

Here’s how this CSS methodology might get applied to a login form:

```
<form class="LoginForm LoginForm--errors">
    <label class="LoginForm-username is-required">
        Username <input type="text" name="username" />
    </label>
    <label class="LoginForm-password">
        Password <input type="password" name="password" />
    </label>
    <button class="LoginForm-button is-inactive">Sign in</button>
</form>
```


## Elementary CSS

My own methodology for managing CSS at scale shares many of the principles and ideas you can find in OOCSS, BEM, SMACSS, SUIT CSS, and other CSS methodologies. But it is intended to be simpler. There are fewer rules to remember, and the class-naming convention is more intuitive.

The process of developing a new web design is broken up into four phases:

- Layout
- Elements
- Widgets
- Modifiers

First, an empty wireframe-like layout is created from a series of sectioning elements such as <section> and <div>:

```
<div class="CONTAINER">
    <header class="BANNER"></header>
    <nav class="NAVIGATION_PRIMARY"></nav>
    <nav class="NAVIGATION_SECONDARY"></nav>
    <main class="MAIN"></main>
    <aside class="ADVERTS"></aside>
    <footer class="FOOTER">
        <nav class="SITEMAP"></nav>
        <div class="LEGAL"></div>
    </footer>
</div>
```

Second, we establish the default presentation for HTML elements that will encapsulate content and render interactive controls. Examples include headings (``h1``, ``h2``, ``h3``), paragraphs (``p``), lists (``ul`` and ``ol``), tables, forms, and so forth.

Third, we identify recurring patterns of content in our designs. We bundle these recurring patterns up into independent modules of markup and styling (and scripting if needed).

These bundles of recurring content patterns I like to call UI “widgets”. For example, here’s the markup for two widgets that render a navigation bar and search form:

```
<div class="NavBar">
    <ul>
        <li><a href="./">Home</a></li>
        <li><a href="about.html">About</a></li>
        <li><a href="learn/">Learn</a></li>
        <li><a href="extend/">Extend</a></li>
        <li><a href="share/">Share</a></li>
    </ul>
</div>
<div class="SearchBox">
    <form action="search.html" method="get">
        <label for="input-search">Search</label>
        <input name="q" type="search" id="input-search" />
        <button type="submit">Search</button>
    </form>
</div>
```

Content — in the form of widgets and naked HTML elements — is then placed within the layout.

Finally, modifier classes are added to vary the default presentation of things.

In the following example the navbar-primary modifier varies the default presentation of the NavBarwidget, and the navbar-selected modifier targets the current selection:

```
<div class="NavBar navbar-primary">
    <ul>
        <li><a href="./">Home</a></li>
        <li><a href="about.html" class="navbar-selected">About</a></li>
        <li><a href="learn/">Learn</a></li>
        <li><a href="extend/">Extend</a></li>
        <li><a href="share/">Share</a></li>
    </ul>
</div>
```

In this methodology, each class performs one, and only one, of the following roles:

- It represents the name of a layout section
- It represents the namespace for a widget
- It modifies the default presentation of something

The three types of classes adopt different naming conventions, to easily tell them apart.

- Layout: UPPER_CASE
- Widget: CamelCase
- Modifier: lower_case

What’s nice about this naming convention is the hierarchy of classes is represented by their letter-casing.

Layout classes, written using all-upper-case letters, scream out from the markup. Examples: NAVIGATION, SIDEBAR, FOOTER.

The CamelCase letter-casing of widgets, which will always be nested within layout classes, are a little less prominent. Examples: MainMenu, ImageGrid, BlogPost.

Modifier classes — which are arguably the least important classes since they merely modify things and are not critical to achieving a default presentation — are the least conspicuous of all, being written in all-lower-case letters. Examples: is-highlighted, has-errors, hidden.

***

It is perfectly acceptable to use long chains of modifier classes. As long as each modifier targets a limited number of properties, these methodology is perfectly sustainable:

```
<nav class="PrimaryNav is-open featured-img"></nav>
```



## Conclusion

All CSS methodologies tackle the scalability and maintainability problem in CSS by providing a class-based system for breaking up big web designs into lots of small, modular, discrete units. Each UI module can be reused over and over throughout a design, and even ported from one project to another if two projects share the same CSS methodology.

In the process, CSS methodologies do much more than fix the CSS scalability problem. They make it easier to develop and iterate a design. They make front-end code easier to read and understand, provide ready-made documentation, and make it easier for multiple people to collaborate on a design.

Adopting a CSS methodology can reduce the learning curve for new designers joining a project, and make for a smoother transition when a project is handed over to a new team.

And because CSS methodologies encourage reuse of existing code, they enforce consistency in visual designs and reduce page size and increase page rendering speed.

CSS methodologies have different class-naming conventions and they carve up web designs along slightly different lines. But the specifics of any particular methodology are less important than the general solutions they provide for modularizing front-end code and making CSS easier to scale.

You can take away the ideas and develop your own categories of classes and devise your own class-naming conventions that work best for you. That’s what I do. Treat the well-known CSS methodologies as a starting point, but every project is different, and I always tweak and extend my CSS methodology to better fit the skill set and creative temperament of the team I’m working with.

## Other CSS methodologies

- [Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/): a design system that uses chemistry terms and analogies (atoms, molecules, and organisms).
- [DoCSSa](http://docssa.info/): a Sass-based CSS methodology by Matthieu Larcher and Fabien Zibi. (offline)
- [csstyle](http://www.csstyle.io/): a Sass/PostCSS-based methodology by Dave Geddes.

**Comment by kieranpotts** [comment]
# CSS Architecture

The Nero UI project aspires to provide a highly structured CSS codebase where the leaky style cascade is controlled. That is done by composing user interfaces from lots of small, autonomous, highly specialised modules, each encapsulated within clear and meaningful class names. This is called CSS architecture. This is how we implement CSS architecture in the Nero UI project.

## CSS Primer

As the name suggests, Cascading Style Sheets set styles in a cascading fashion. Multiple style rules may be applicable to any one element in the HTML markup. Which rules get applied is determined by cascading down through every ruleset in every style sheet looking for matches. The more specific rulesets override the more general ones. In addition, some applied styles will cascade down through the DOM and apply themselves to all of an element's descendants, too.

This inheritance model is intrinsic to CSS and cannot be turned off.

In simple, small-scale user interfaces the style cascade can be hugely beneficial. We can set a property just once and have it trickle through to every relevant part of our design. The result: small file sizes, efficient rendering and highly consistent user interfaces.

But the style cascade can also be a bit of nuisance. The CSS inheritance model tightly couples the presentation of every component of a web page, and that makes it difficult to change stuff. Every change has the potential to inadvertently alter the presentation of other things. We end up frequently resetting bunches of inherited properties that we don't want, which rather negates the benefits of the cascade.

The bigger the project, the more the style cascade is a hindrance than a help. It's why CSS is notoriously difficult to maintain in high-fidelity designs (where there is lots of detail and variation in the presentation of things), in long-running projects with many collaborators, and in projects where the team members have widely differing specialities and abilities or where there is a high staff turnover. In these scenarios if everyone is not on the same page in terms of how to structure and manage CSS, a codebase can rapidly snowball into an unmanageable mess.

CSS does not scale at all well for today's richly formatted, rapidly iterated, interactive and dynamic web application interfaces.

Extended CSS languages – Sass, Less, Stylus, et al – make things a bit better. And some of their best ideas – like variables – are getting ported to native CSS (slowly). But extended CSS languages don't truly fix the scalability issue. They just skirt around the problem by encouraging code duplication, deep nesting of selectors and other baddies.

What we really need is a proper scoping mechanism, a standardised way to encapsulate styles within fragments of markup. This would allow us to split up complex designs into lots of bite-size modules of HTML, CSS and optionally JavaScript. Each module could be developed and maintained separately of the rest. Team members could more easily work on different things in parallel. And it would be easier to take UI components from previous projects and plug them straight in to new projects, speeding up development cycles.

Modular CSS would be the ultimate scalable CSS.

Things are moving in this direction. The Shadow DOM – part of the Web Components standard dreamt up at the W3C – encapsulates styles by default. There were also proposals for an ``@scope`` rule and a ``scoped`` attribute for the ``<style>`` element, though take up of these ideas has stalled. Some browser vendors have implemented the new ``contain`` and ``all`` properties, which scope styles within a container and reset inherited styles respectively.

```css
.scoped {
  contain: strict;
}

.reset { 
  all: unset; 
}
```

None of these things are useful right now, because they do not yet work consistently in all mainstream browsers.

No matter. Things are certainly moving in the direction of modularity. In the future we will have standardised methods of scoping presentational styles to particular bits of an HTML document and locking out the inheritance chain from others. This will make it possible to develop web interfaces as a series of stand-alone UI components. And that will make it much easier to manage the codebases of large-scale web projects.

In the meantime we must find a way to do this ourselves. One option would be to polyfill Web Components in browsers that lack native support (http://webcomponents.org/polyfills/). This is expensive (in terms of page weight) and it makes static rendering dependent upon JavaScript. Bespoke solutions such as CSS Modules (https://github.com/css-modules) have the same problems.

Right now, the best option is to devise our own system to manage CSS at scale.


## Web design methodologies

A web design methodology is a system for applying CSS to HTML. The primary aim is to make it possible to break up big, complex web designs into lots of smaller, simpler modules of HTML and CSS (and optionally JavaScript) that can be developed and maintained independently of one another. But a good web design methodology will achieve much more. It will:

- Make the design process faster.
- Make front-end code easier to read and understand.
- Make it easier for multiple people to collaborate on a design.
- Make for a smoother transition in the handover of a project to a new team.
- Reduce the learning curve for new designers joining a project.
- Help developers understand the relationship between HTML and CSS in a project.
- Make it easier to change things.
- Make it easier to scale a design to accommodate more content and functionality.
- Make it easier to add more variety to the visual presentation.
- Make it easier to reuse things that already exist in a project codebase.
- Make it easier to port UI components from a previous project into a new one.
- Encourage production of well-formed, accessible, standards-compliant code.
- Enforce consistency and limit duplication and bloat.
- Reduce page size and increase page rendering speed.
- Provide ready-made documentation.
- Increase productivity.
- And make web design a more rewarding, enjoyable experience.

Web design methodologies benefit all kinds of web projects, big and small. They benefit freelancers who work solo on small-scale web projects. And they are absolutely essential for large, long-running and fast-moving projects.

A handful of front-enders have developed their own web design methodologies, evolving them into formal design systems and published them for everyone's benefit:

- Nicole Sullivan's Object-Oriented CSS Framework (OOCSS), published in 2009 (https://github.com/stubbornella/oocss/), was the first web design methodology to become widely adopted. It is still hugely influential.
- Jeremy Clarke's Don't Repeat Yourself principles (DRY CSS), presented at ConFoo in 2012 (https://vimeo.com/38063798), has similar goals to OOCSS, but offers a very different solution. 
- Brad Frost followed up with a widely circulated blog post titled "Atomic Design" in 2013 (http://bradfrostweb.com/blog/post/atomic-web-design), which got us thinking about web interfaces as a hierarchy of progressively more complex objects built up from lots of smaller ones. 
- Block Element Modifier (BEM, http://bem.info), developed by the dev team at Yandex, is today probably the most popular CSS design system.
- Jonathan Snook's Scalable and Modular Architecture for CSS (SMACSS, http://smacss.com) is well presented.
- Nicolas Gallagher's SUIT CSS (http://suitcss.github.io) combines a class naming system with extended CSS syntax.

There are many others.

Nero CSS implements a web design methodology that is influenced by many of the ideas and conventions of the earlier methdologies listed above. It is probably most similar to BEM, but the class naming convention is different.

The Nero CSS architecture is a pure CSS methodology, though it could easily be adopted for a codebase written in Sass or Less or any other superset of CSS. Many of the features of extended CSS languages become redundant once you've got a solid web design methodology in place.

Nero CSS prescribes strict class naming conventions, which reflect how this design system separates out the various concerns in a user interface into discrete units of code.


## Reset

The first stage of the Nero CSS methdology is to define the default presentation for all of the base HTML elements that you will need to encapsulate content and to render interactive controls. Things that will probably need your attention include:

- Headings, paragraphs, lists and hyperlinks.
- Data tables.
- Forms, fieldsets and legends.
- Input controls and buttons.

Most of the CSS at this stage will be made up of plain old type selectors.

```css
h2 {
  font-size: 2rem;
  line-height: 1.2;
  margin-top: 2em;
  margin-bottom: 1em;
}

h3 {
  font-size: 1.4rem;
  line-height: 1.2;
  margin-top: 2em;
  margin-bottom: 1em;
}

p {
  font-size: 1.4rem;
  line-height: 1.2;
  margin-bottom: 1em;
}
```

Though it is easier to deal with the presentation of one element at a time, you might consider reorganising your selectors around properties rather than elements, to reduce duplication of the same properties and values.

```css
h2        { font-size: 2rem; }
h3, p     { font-size: 1.4rem; }

h2, h3, p { line-height: 1.2; }

h2, h3    { margin-top: 2em; }
h2, h3, p { margin-bottom: 1em; }
```

For efficiency some inheritable typographic styles may be set on the <body> element and allowed to cascade from there.

```css
body {
  color: hsl(0, 0%, 27%);
  font-family: Roboto, Arial, Helvetica, sans-serif;
  font-size: 62.5%;
}
```

You will probably want to reset some of the default properties set by the browser, to make a more consistent cross-browser base from which to build out your design.

```css
* {
  margin: 0;
  padding: 0;
}

html   { font-size: 62.5%; }
html * { font-size: 100%; }
body * { font-family: inherit; }
```

As well as element types, you might want to target pseudo-classes, pseudo-elements and element attributes, too.

```css
a:hover { opacity: 1; }
q::before { content: ' «'; }
button[disabled] { opacity: 0.5; }
```

But descendant selectors, child selectors and sibling selectors should be used more sparingly. The default presentation of individual elements should not be dependent upon those elements being used in any particular context.

```css
li a {}
li > a {}

h2 + h3 {}
h2 ~ p {}
```

Initially, prefer to promote consistency by fleshing out an exhaustive suite of default styles for individual HTML elements. Later, when you make your utilities and components, if you find yourself resetting lots of inherited properties, consider iterating towards a lower baseline of styling, reducing the number of properties applied directly to elements via their type selectors and increasing your reliance on classes to build up more intricate designs.

There is something to be said for having a very minimalist baseline design. I have an elements style sheet that I use as the "reset" for every web project. This baseline design rarely changes from one project to the other. Because every project is built using the same web design methodology and on the same design foundation, I can easily port UI components – things like tabbed interfaces and modal popups and other such components – from one project to another.


## Layout

The second stage of the Nero CSS methodology is to give your HTML documents a structure. Use a series of sectioning elements – ``<main>``, ``<header>``, ``<footer>``, ``<article>``, ``<aside>``, ``<address>``, ``<nav>`` and ``<section>``, and also generic ``<div>`` elements – to establish sections in which content will be later placed. Give each section a unique class name, written in full upper case with words delimited by underscores.

```css
<body>
    <div class="CONTAINER">
        <header class="BANNER"></header>
        <nav class="NAVIGATION_PRIMARY"></nav>
        <nav class="NAVIGATION_SECONDARY"></nav>
        <main class="MAIN"></main>
        <aside class="ADVERTS"></aside>
        <footer class="FOOTER">
            <nav class="SITEMAP"></nav>
            <div class="LEGAL"></div>
        </footer>
    </div>
</body>
```

In the CSS, use class selectors to target each section. Do not qualify the class selectors with types. That will give you the freedom to swap around sectioning elements in the markup – for example, switching an ``<aside>`` for a generic ``<div>``, so changing the semantics of the document – without needing to refactor your style sheets, too.

```css
.CONTAINER {}
.BANNER {}
.NAVIGATION_PRIMARY {}
.NAVIGATION_SECONDARY {}
.MAIN {}
.ADVERTS {}
.FOOTER {}
.SIDEBAR {}
.LEGAL {}
```

In mobile-optimised views, document sections will be stacked vertically, one on top of the other. But in larger viewports sections may be positioned to create a layout.

The layout rulesets are concerned exclusively with establishing a basic visual structure in which to place content (which will be styled separately). Layout sections may be positioned and they may be given backgrounds, borders, padding and margins. But layout classes must never set any typographic styles, or alter the presentation of form controls, or try to influence the presentation of anything that may be placed within any section of the layout. The aim is to have the freedom to move content around in the layout without needing to make changes to the CSS.

To this end, do not set any inheritable properties via layout classes. Apply only non-inheritable properties to layout sections. These will not cascade down and affect nested content. Non-inheritable properties include:

- ``display``
- ``float``
- ``position``
- ``top``, ``bottom``, ``left``, ``right``
- ``width``, ``height``, ``max-width``, etc.
- ``background``, ``background-color``, ``background-image``, etc.
- ``border``, ``border-color``, etc.
- ``margin``, ``margin-left``, etc.
- ``padding``, ``padding-left``, etc.
- ``z-index``

You might not want every page of a web site to follow the exact same layout. So you can modify and extend the default layout for different pages. Where a layout differs from the default, give the root ``<html>`` element a unique class name, using the same naming convention as for layout sections. Example:

```html
<html class="SEARCH">
```

Now you can use the "SEARCH" class to encapsulate aspects of the layout that are specific to your web site's search page. You might make adjustments to sections that already exist in the default layout. And you might introduce entirely new layout sections that exist only on the search page. In the following example there is one selector for modifying and extending the default presentation of the global ``NAVIGATION_PRIMARY`` section, and there are two selectors for new sections that exist only on the search page: ``SIDEBAR`` and ``PAGINATION``.

```css
.CONTAINER {}
.BANNER {}
.NAVIGATION_PRIMARY {}
.NAVIGATION_SECONDARY {}
.MAIN {}
.ADVERTS {}
.FOOTER {}
.SITEMAP {}
.LEGAL {}

html.SEARCH .NAVIGATION_PRIMARY {}
html.SEARCH .SIDEBAR {}
html.SEARCH .PAGINATION {}
```

Notice that the ``SEARCH`` class is qualified with a type selector. This just makes it clearer that it is the encapsulating class for a whole template and not just for one section of the layout.

In designs with lots of layout variations you might consider prefixing global layout sections with ``GLOBAL_``, to tell them apart from sections that exist only in certain templates.

```css
.GLOBAL_CONTAINER {}
.GLOBAL_BANNER {}
.GLOBAL_NAVIGATION_PRIMARY {}
.GLOBAL_NAVIGATION_SECONDARY {}
.GLOBAL_MAIN {}
.GLOBAL_ADVERTS {}
.GLOBAL_FOOTER {}
.GLOBAL_SITEMAP {}
.GLOBAL_LEGAL {}

html.SEARCH .GLOBAL_NAVIGATION_PRIMARY {}
html.SEARCH .SIDEBAR {}
html.SEARCH .PAGINATION {}
```

Alternatively, you might use a grid system to create the various sections in which content will be laid out.

```html
<div class="GRID">
    <header class="GRID_ROW">
        <div class="GRID_COLUMN_ONE_OF_TWO">
            ...
        </div>
        <div class="GRID_COLUMN_ONE_OF_TWO">
            ...
        </div>
    </header>
    <main class="GRID_ROW">
        <div class="GRID_COLUMN_TWO_OF_FOUR">
            ...
        </div>
        <div class="GRID_COLUMN_ONE_OF_FOUR">
            ...
        </div>
        <div class="GRID_COLUMN_ONE_OF_FOUR">
            ...
        </div>
    </main>
    <footer class="GRID_ROW">
        <div class="GRID_COLUMN_ONE_OF_ONE">
            ...
        </div>
    </footer>
</div>
```

Personally, I find traditional layouts much more flexible. Because each section is given a unique name, different sections can collapse to mobile-friendly layout at different breakpoints. This is difficult to achieve – though not impossible – with generic grid systems. I do use grid systems but not for primary layout. Instead I use grids to set out content within an existing layout section. I will cover this technique in the next chapter.

#### Layout modifiers

Modifiers may be used also to tweak the default presentation of sections of the layout, too. Layout modifiers should use the same naming convention as for component modifiers, taking the name of their parent layout section (written lower case) as a prefix.

```html
<header class="BANNER banner-homepage">
```

In the CSS, layout modifiers should be further qualified against the class name of the layout section that they modify. This just locks things down that little bit more, and it's not going to be such a big issue if a layout modifier is accidentally used in the wrong place.

```css
.BANNER {
    background: hsl(0, 0%, 93%);
    display: table-cell;
    padding: 2rem 0;
    vertical-align: middle;
}

.BANNER.banner-homepage { height: 100vh; }
```


## Components

HTML's extensive suite of native element types still does not provide us with a large enough vocabulary from which we can design rich, modern web interfaces. We often need to group together multiple HTML elements to achieve a particular effect. We might combine an ``<input>`` with a ``<button>`` to create a search box, or repurpose the ``<ul>`` element to make a navigation menu.

We could just plonk these things down directly inside the sections of the layout where we want them to be.

```html
<nav class="NAVIGATION">
    <ul>
        <li><a href="/">Home</a></li>
        <li><a href="/about">About</a></li>
        <li><a href="/learn">Learn</a></li>
        <li><a href="/extend">Extend</a></li>
        <li><a href="/share">Share</a></li>
    </ul>
    <form action="search.html" method="get">
        <label for="input-search">Search</label>
        <input name="q" type="search" id="input-search" />
        <button type="submit">Search</button>
    </form>
</nav>
```

We would then alter the presentation of individual elements when they appear within a particular layout section by using the name of that section as a target for our CSS selectors.

```css
.NAVIGATION ul {}
.NAVIGATION li {}
.NAVIGATION a {}
.NAVIGATION form {}
.NAVIGATION label {}
.NAVIGATION input {}
.NAVIGATION button {}
```

The big disadvantage of this approach is that the presentation of content and interactive controls – in this case, a navigation menu and search box – is tightly coupled to its location within the layout. If we move the content, its presentation breaks. And we can't easily reuse existing markup elsewhere. More work for us.

A better approach is to identify recurring patterns of content in our designs – in Nero CSS we call these components, they are also commonly known simply as blocks or patterns – and to bundle them up into independent modules of markup and styling (and maybe scripting) that can be freely moved around and reused in different contexts.

A component is any reusable user interface component. It could be something simple like a search box or a navigation menu. Or it could be something more complex and dynamic like a modal popup, a photo slideshow or a real-time user comment feed. Some components may be used on every page of a web site, while others may be rolled out only occasionally.

So the next stage of the methodology is to design all of the components that you will need to render your content and interactive controls. Creating components is easy. Simply encapsulate the markup for each component in a unique class. Components use the ``CamelCase`` naming convention, to distinguish them from layout sections (``UPPER_CASE``) and utility classes (``lower-case``).

Below, the content for our navigation menu is now encapsulated in a new ``<div>`` which is given the class ``NavBar``. And the markup for our search box is now encapsulated in another ``<div>`` which is given the class ``SearchBox``.

```html
<div class="NavBar">
    <ul>
        <li><a href="/">Home</a></li>
        <li><a href="/about">About</a></li>
        <li><a href="/learn">Learn</a></li>
        <li><a href="/extend">Extend</a></li>
        <li><a href="/share">Share</a></li>
    </ul>
</div>

<div class="SearchBox">
    <form action="search.html" method="get">
        <label for="input-search">Search</label>
        <input name="q" type="search" id="input-search" />
        <button type="submit">Search</button>
    </form>
</div>
```

Here's the full and final markup, with the ``NavBar`` and ``SearchBox`` components nested inside the original ``NAVIGATION`` layout section.

```html
<nav class="NAVIGATION">
    <div class="NavBar">
        <ul>
            <li><a href="/">Home</a></li>
            <li><a href="/about">About</a></li>
            <li><a href="/learn">Learn</a></li>
            <li><a href="/extend">Extend</a></li>
            <li><a href="/share">Share</a></li>
        </ul>
    </div>
    <div class="SearchBox">
        <form action="search.html" method="get">
            <label for="input-search">Search</label>
            <input name="q" type="search" id="input-search" />
            <button type="submit">Search</button>
        </form>
    </div>
</nav>
```

Each component's encapsulating class acts as a sort of namespace for presentational styles that are unique to the component. We no longer need to use the name of a parent layout section as a namespace.

```css
.NavBar {}
.NavBar ul {}
.NavBar li {}
.NavBar a {}
.NavBar a:hover {}

.SearchBox {}
.SearchBox form {}
.SearchBox label {}
.SearchBox input {}
.SearchBox button {}
```

Now, the ``NavBar`` and ``SearchBox`` components will look the same no matter where we put them. Their presentation is not dependent upon their markup being located anywhere particular. We can move them about in the layout. We can duplicate them. And we can do these things without touching the CSS.

Components should be designed to fit 100% of the width of whatever container we put them in. If you want to float a component to the left or right side of its container, or if you want to apply fixed or absolute positioning to a component, better to wrap the component in a new layout section and position that instead. Keep these concerns – structure and content – separate.

Avoid using semantic sectioning elements – ``<main>``, ``<header>``, ``<footer>``, ``<article>``, ``<aside>``, ``<address>``, ``<nav>`` and ``<section>`` – either within or to encapsulate components. These elements should be reserved for layout sections. Components may be contained in generic divider (``<div>``) elements or any other appropriate block-level element such as a ``<ul>``, ``<table>`` or even a ``<blockquote>``.

```html
<table class="MiniCalendar">
    <thead>
        <tr class="nav">
            <th class="prev"><a href="?m=02&y=2015">Prev</a></th>
            <th class="this" colspan="5">January 2015</th>
            <th class="next"><a href="?m=02&y=2015">Next</a></th>
        </tr>
        <tr class="days">
            <th>Mon</th>
            <th>Tue</th>
            <th>Wed</th>
            <th>Thu</th>
            <th>Fri</th>
            <th>Sat</th>
            <th>Sun</th>
        </tr>
    </thead>
    <tbody>
        ...
    </tbody>
</table>
```

In practice it will not always be possible to make every component entirely location agnostic. Some components may be positioned in a layout section that has a dark background, in which case the colours of the component will need to be adjusted accordingly. And responsive breakpoints can be an issue. Breakpoints can be set only according to the size of the viewport or display, and not according to the size of any particular element in the document. Thus the breakpoints that we set for components will be dependent upon those components being located in a container of a certain size. These are not problems for global components, which appear in the same spot on every page: the web site's logo, navigation bar, footer links, etc. But these are problems that are worth trying to avoid in components that are meant to be reusable in lots of different contexts. In particular, try to avoid relying on media queries to adjust the layout of things within a component. Doing so increases the coupling between components and layouts, as the components are required to be located in certain sections of the layout.

Components work best when they are quite abstract. A ``ComparisonTable`` component would be preferable to a ``PriceComparison`` component, since the former could be reused to compare things besides prices. The more abstract the component, the greater the opportunity for reusing it for different things.

It is recommended that all content be encapsulated in a component of some sort. In theory, content can be marked up as a mix of components and stand-alone elements. Consider the following example. There is one component, ``ArticleHeadline``, which encapsulates an ``<h1>`` and ``<p>`` that represent the main headline and trailer text for an article. But the main copy and opening image of the article is placed directly inside the ``MAIN`` layout section. This content will be styled exclusively with the default properties set for the ``<img>`` and ``<p>`` element types. (Actually, the presentation of the ``<img>`` element is extended by application of a global utility class, "float-right". We'll discuss utility classes in the next section.)

```html
<main class="MAIN">
    <header class="HEADLINE">
        <div class="ArticleHeadline">
            <h1>Headline...</h1>
            <p>Strapline...</p>
        </div>
    </header>
    <img src="picture.png" alt="" class="float-right" />
    <p>Lorem ipsum...</p>
</main>
```

This can work well, but in general I find it easier if all visible content and every interactive control is encapsulated in a dedicated component. It means that when I want to change the design of something I can work exclusively with components; I do not have to change the default presentation of individual HTML elements, which will likely have wider consequences. Here's the same content as before but now the main body of the article is encapsulated in a new component called ``ArticleBody``:

```html
<main class="MAIN">
    <header class="HEADLINE">
        <div class="ArticleHeadline">
            <h1>Headline...</h1>
            <p>Strapline...</p>
        </div>
    </header>
    <div class="ArticleBody">
        <img src="picture.png" alt="" class="float-right" />
        <p>Lorem ipsum...</p>
    </div>
</main>
```

Components will, of course, inherit the default properties of the element types that are included in their markup. Those default styles are then changed and extended on a component-by-component basis. Sometimes you might want to reset some inherited properties. For example you might want to change the default color and text-decoration values for ``<a>`` elements when they are used in the ``NavBar`` component. A few little overrides like this are perfectly acceptable. But if you catch yourself doing lots and lots of resets, consider applying fewer default properties to elements via non-qualified type selectors, so that less stuff is inherited by components.

Web interfaces built using the Nero CSS methodology will be composed from lots and lots of components. You might even have components nested inside other components. Consider the following markup:

```html
<div class="Dialog">
    <div class="DialogHeader">
        <h2>Popup title</h2>
    </div>
    <div class="DialogBody">
        <p>Popup message</p>
    </div>
    <div class="DialogFooter">
        <p><small>Optional footer content</small></p>
    </div>
</div>
```

This is the markup for the content of a ``Dialog`` component, which pops up to present the user with critical information or to demand mandatory user input. The markup in this example is simplified, but the structure of popup dialogs can grow to become pretty complex. So the HTML author has opted to break up the content of the ``Dialog`` component into three separate components, each specialising in presenting the content of a distinct sub-section. Those child components are called ``DialogHeader``, ``DialogBody`` and ``DialogFooter``. Here are the selectors:

```css
.Dialog {}
.DialogHeader {}
.DialogHeader h2 {}
.DialogBody {}
.DialogBody p {}
.DialogFooter {}
.DialogFooter p {}
```

Because the sub-components are encapsulated in their own namespaces, each can be developed, tested and maintained separately of the other sub-components. And each component specialises in just one thing. The parent ``Dialog`` component would be responsible only for wrapping everything up in a box and positioning it on the page. Each of the other three components need only to focus on presenting the content in one section of a dialog window.

The components ``Dialog``, ``DialogHeader``, ``DialogBody`` and ``DialogFooter`` are clearly related. But there's nothing stopping us from combining entirely unrelated components to compose complex UI designs. Here's the markup for a tabbed interface that is made from lots of different things:

```html
<div class="Tabs">
    <ul>
        <li><a href="#panel1" class="tabs-selected">Popular</a></li>
        <li><a href="#panel2">Just in</a></li>
        <li><a href="#panel3">Editor's choice</a></li>
    </ul>
</div>

<div class="Panel" id="panel1">
    <h2>Most popular stories</h2>
    <div class="Slats">
        <div class="Slat">
            <h3><a href="#">This is the title</a></h3>
            <p>Lorem ipsum dolor sit amet...</p>
            <p class="meta">10 January, 2015</p>
        </div>
        <div class="Slat">
            <h3><a href="#">This is the title</a></h3>
            <p>Lorem ipsum dolor sit amet...</p>
            <p class="meta">10 January, 2015</p>
        </div>
        <div class="Slat">
            <h3><a href="#">This is the title</a></h3>
            <p>Lorem ipsum dolor sit amet...</p>
            <p class="meta">10 January, 2015</p>
        </div>
    </div>
</div>
<div class="Panel panel-is-closed" id="panel2">
    ...
</div>
```

In the next example, a grid system (a component) is used to arrange in rows and columns the content of numerous other nested components, all contained within a section of the wider layout called ``MAIN``.

```html
<main class="MAIN">
    <div class="Grid">
        <div class="grid-row">
            <div class="grid-column-one-of-two">
                <div class="Promotion">
                    ...
                </div>
            </div>
            <div class="grid-column-one-of-two">
                <div class="PullQuote">
                    ...
                </div>
            </div>
        </div>
        <div class="grid-row">
            <div class="grid-column-two-of-four">
                <div class="Box box-shade1">
                    ...
                </div>
            </div>
            <div class="grid-column-one-of-four">
                <div class="Box box-shade2">
                    ...
                </div>
            </div>
            <div class="grid-column-one-of-four">
                <div class="Box box-shade3">
                    ...
                </div>
            </div>
        </div>
    </div>
</main>
```

Another methodology for composing complex UIs is to have components extend from other components. Here's the markup for a component called ``Box``. It's role is simple: to present a short message in a box. There is an optional button that when clicked will hide the box.

```html
<div class="Box">
    <h2>Message</h2>
    <p>Description</p>
    <button>Action</button>
</div>
```

Here are the selectors we'd need to style this component:

```css
.Box {}
.Box h2 {}
.Box p {}
.Box button {}
```

Now imagine that we wanted to modify the default presentation of boxes when the message is particularly important or requires feedback from the user. One option would be to add a utility class (this technique is covered in the next section):

```html
<div class="Box box-alert">
    <h2>Message</h2>
    <p>Description</p>
    <button>Action</button>
</div>
```

Another option is to create a new component that inherits the default presentation of the Box component. Here's how that can be done:

```html
<div class="Box">
    <h2>Message</h2>
    <p>Description</p>
    <button>Action</button>
</div>
<div class="Box BoxAlert">
    <h2>Message</h2>
    <p>Description</p>
    <button>Action</button>
</div>
```

In the markup, notice that the ``BoxAlert`` component (the second one) also uses the encapsulation class for the ``Box`` component. Thus it will inherit the styles for the ``Box`` component and then modify and extend those itself.

Component inheritance works great when two or more components share a similar HTML structure. You might go as far as deliberately making abstract components that are not intended to be used on their own, but which provide a basic foundation from which to build other components.

Component inheritance can be convenient, but it can also get messy. Deep levels of inheritance tend to make things harder to maintain, not less so. Try to stick with just a single tier of inheritance. And prefer to err on the side of composition over inheritance. It's generally much easier to compose complex interfaces by combining lots of smaller components with no dependencies of their own.


#### Component modifiers

The default presentation of components can be tweaked and extended by application of additional modifier classes. Component modifiers should be prefixed with the name of the component that they modify, converted to lower case.

```css
.Dialog {
    border: 1px solid hsl(0, 0%, 60%);
    left: 50%;
    margin-left: -400px;
    margin-top: -40px;
    padding: 2rem;
    position: absolute;
    top: 50%;
    width: 800px;
}

.Dialog.dialog-alert { border-color: hsl(15, 100%, 50%); }
```

Usage:

```html
<div class="Dialog dialog-alert">
```

Component modifiers may change the default presentation of the component's container element, or of any nested elements that form the component's structure. In the following example the border-color property will apply only to the container element for each Dialog instance while the font-weight property will be inherited by all of the component's descendant elements.

```css
.Dialog.dialog-alert {
    border-color: hsl(15, 100%, 50%);
    font-weight: bold;
}
```

Modifiers may be needed to target specific instances of an element within a component structure, too.

```html
<div class="Dialog dialog-alert">
    <div class="dialog-header>
        <h2>...</h2>
        <p>...</p>
    </div>
    <div class="dialog-body>
        ...
    </div>
    <div class="dialog-footer>
        <button>...</button>
    </div>
</div>
```

Here are the selectors we might use to target each element that composes this component:

```css
.Dialog {}
.Dialog .dialog-header {}
.Dialog .dialog-header h2 {}
.Dialog .dialog-header p {}
.Dialog .dialog-body {}
.Dialog .dialog-footer {}
.Dialog .dialog-footer button {}
```

The prefix convention for component modifiers is important. The point is to reduce the chances of the same class name being used for two unrelated things, which might cause unexpected property leakages. Consider the following example.

```html
<div class="Headline">
    <p class="tagline">Tagline</p>
    <h1>Headline</h1>
    <p class="teaser">Teaser</p>
    <p class="meta">Author, Date</p>
</div>
```

Here are the selectors we'd need to style Headline component instances:

```css
.Headline {}
.Headline h1 {}
.Headline p {}
.Headline p.tagline {}
.Headline p.teaser {}
.Headline p.meta {}
```

There are three modifier classes here: "tagline", "teaser" and "meta". The problem with these modifiers is that there is some risk of cross-contamination. It is entirely possible that the following selectors may exist some place else in our style sheets.

```css
p.tagline {}
p.teaser {}
p.meta {}
```

Or:

```css
*.tagline {}
*.teaser {}
*.meta {}
```

Any properties set by these selectors would be inherited by the Headline component. The purpose of prefixing component modifiers is to avoid exactly this scenario.

```html
<div class="Headline">
    <p class="headline-tagline">Tagline</p>
    <h1>Headline</h1>
    <p class="headline-teaser">Teaser</p>
    <p class="headline-meta">Author, Date</p>
</div>
```

Here are the final selectors:

```css
.Headline {} 
.Headline h1 {}
.Headline p {}
.Headline p.headline-tagline {}
.Headline p.headline-teaser {}
.Headline p.headline-meta {}
```

Now there is much less risk that the presentation of the Headline component will be contaminated by other things. This naming convention has another benefit: it makes it easier to see in the markup which modifiers are global and which belong to particular components.

We could now optimise some of our selectors for the Headline component. After all, the "headline-" prefix used for this component's modifier classes provides adequate encapsulation on its own, does it not?

```css
.Headline {}
.Headline h1 {}
.Headline p {}

p.headline-tagline {}
p.headline-teaser {}
p.headline-meta {}
```

Personally, I prefer to always include the parent component class when selecting child classes.

```css
.Headline {}
.Headline h1 {}
.Headline p {}
.Headline p.headline-tagline {}
.Headline p.headline-teaser {}
.Headline p.headline-meta {}
```

This is clearer, consistent and a little more robust. And this is how it would be compiled by a preprocessor if you were using an extended CSS syntax to nest a component's selectors.

```css
.Headline {

    h1 {}

    p {
        &.headline-tagline {}
        &.headline-teaser {}
        &.headline-meta {}
    }

}
```

But try to avoid combinator selectors that are lengthier than this. Selectors should not need more than one descendant level. Prefer to create new classes than to found selectors on the structure of a component's markup. Things like immediate child selectors and pseudo classes (e.g. .Headline > h1 and :first-of-type) are best avoided, too. They create a tight coupling between HTML and CSS, making it that bit harder to extend and modify components in the future.

#### Dynamic modifiers

Components may be progressively enhanced with dynamic behaviours. Imagine a component called Slideshow that displays a series of images. By default, when the HTML is initially rendered, all of the images in a slideshow might be visible, presented one after the other in a vertical layout. The markup might look something like this:

```html
<div class="Slideshow">
    <a href="sydney.html">
        <img src="sydney.png" alt="Sydney" />
    </a>
    <a href="melbourne.html">
        <img src="melbourne.png" alt="Melbourne" />
    </a>
    <a href="perth.html">
        <img src="perth.png" alt="Perth" />
    </a>
    <a href="adelaide.html">
        <img src="adelaide.png" alt="Adelaide" />
    </a>
    <a href="darwin.html">
        <img src="darwin.png" alt="Darwin" />
    </a>
</div>
```


Then a JavaScript program called Slideshow.js could come along and dynamically add some classes, which the CSS uses to restyle and animate the slideshow. This is what the markup might look like at one moment in time:

```html
<div class="Slideshow slideshow-is-animating">
    <a href="sydney.html" class="slideshow-is-hidden">
        <img src="sydney.png" alt="Sydney" />
    </a>
    <a href="melbourne.html" class="slideshow-is-visible">
        <img src="melbourne.png" alt="Melbourne" />
    </a>
    <a href="perth.html" class="slideshow-is-hidden">
        <img src="perth.png" alt="Perth" />
    </a>
    <a href="adelaide.html" class="slideshow-is-hidden">
        <img src="adelaide.png" alt="Adelaide" />
    </a>
    <a href="darwin.html" class="slideshow-is-hidden">
        <img src="darwin.png" alt="Darwin" />
    </a>
</div>
```

Dynamic modifiers – any class that is dynamically switched in and out of a web page by a client-side JavaScript program – should be prefixed with "is-" followed by a verb or noun that describes the component's current state. Temporary classes used to transition something from one state to another are named "is-" followed by a verb. Permanent classes that fix the state of something are named "is-" followed by an adjective.

```css
.is-opening {}
.is-open {}
.is-closing {}
.is-closed {}
```

Dynamic modifiers that belong to a particular component should be further prefixed with the component's name (written lower case) and encapsulated in the component's namespace.

```css
.Slideshow.slideshow-is-animating {}
.Slideshow a.slideshow-is-visible {}
.Slideshow a.slideshow-is-hidden {}
```

Classes added to a document for the purpose of feature detection should be prefixed with "supports-".

```css
.supports-cssboxsizing {}
.supports-csstransforms3d {}
.supports-webworkers {}
```

The popular Modernizr library can be configured to add a custom prefix to the classes that it injects into documents. Use the "className prefix" option when you download the library from here: http://modernizr.com/download/

Classes are often used as targets for DOM queries, too:

```js
document.querySelectorAll( '.popup' );
```

Personally, I prefer to use custom data-* attributes for this purpose. Using data-* attributes – and <data> elements, sometimes – as the exclusive interface between HTML and JavaScript leaves class attributes as the primary interface between HTML and CSS. This is a convenient dichotomy, and it is one less concern for classes.

```js
document.querySelectorAll( '[data-popup="true"]' );
```


## Utilties

When something needs to be presented slightly differently than normal, give it a utility class. Utilities are the fourth and final component of the Nero CSS methodology.

Utility classes are written full lower case with words delimited by hyphens: ``is-required``, ``sidebar-is-hidden`` and ``contextmenu-selected``, for example. The first letter must be an alphanumeric character; hyphen prefixes (``-util``) are illegal, as they are reserved for browser vendors.

Utility classes should be highly specialised, each setting just a few narrow properties.

As a general (but not absolute) rule, it should not be necessary to apply utilities to achieve a default presentation. Utilities should be optional. They should only tweak and extend the default presentation of something. A document stripped of all its utility classes should render perfectly well. There are a few special cases; you might need to apply the clearfix hack, via a utility called ``clearfix``, to achieve the desired default presentation, for example.

Try to avoid using utilities to reset properties inherited from an element's defaults. Ideally, both components and utilities should progressively enhance the default presentation of HTML elements, not regressively degrade them. There is obviously a balance to be found. Set too many properties directly to naked element types, and you might find yourself needing to unset lots of inherited properties when elements are used in special contexts. Set too few default properties, and you will more likely need to maintain a large battery of components and utilities to build out from that minimalist baseline design.

High-fidelity designs – where there is lots of detail and variation in the presentation of things – will inevitably end up with lots of utilities littered about the HTML. That's okay. The convenience of utilities in such designs makes up for any difficulties in maintaining neat and tidy markup.

Utilties can be used in lots of different contexts. A utilities could be:

- Global.
- Specific to one or more element types.
- Specific to a component.
- Specific to certain elements within a component.
- Specific to a section of the layout.

And some of those utilities may be dynamic, swapped in and out of the markup by clientside scripting.

#### Global utilities

Global utilities are specialist classes that are intended to be used almost anywhere. State your intention and include the universal selector (*) when defining global utilities.

```css
*.error { color: red; }
```

Be judicious in your use of global utilities. They are inherently leaky and may get added to things for which they were not designed, possibly leading to unexpected presentational effects. And rulesets for naked elements and components will often have higher specificity, overriding the styles of global utilities when they are applied to the same elements and components. And that in turn may encourage use of the nasty !important keyword to give the properties of global utilities greater precedence. This keyword ought to be used only where you want to ensure that a property cannot be overridden by the client's own style sheets.

```css
*.error { color: red !important; }
```

For these reasons, wherever practical prefer to remove utilities from the global space and instead qualify to what element types, layout sections or components they can be applied. This will boost the specificity of the utilities, eliminating the need for the !important keyword and reducing the likelihood of mis-use of the classes.

```css
p.error, ul.error,
input.error {
    color: red;
}
```

That's not to say that global utilities are not useful. They certainly are useful for general utilities like the clearfix hack and for applying niche presentational effects like drop shadows and rounded corners that may recur on different, unrelated components of an interface.

```html
<div class="Popup clearfix box-shadow rounded-corners">
```

#### Element-specific utilities

HTML's native elements do not always give designers the variety they crave. We often want to introduce variations in the default presentation of individual elements, without consideration to the specific contexts in which those elements are used. Utility classes that are designed to vary the default presentation of HTML elements should be qualified against the relevant types.

```css
p {
    font-size: 1.4rem;
}

p.standout {
    font-size: 1.8rem;
    font-weight: bold;
}

p.smallprint {
    font-size: 1rem;
}
```
